File: src/auth/dev.ts
// src/auth/dev.ts
//
// AUTH_MODE=dev static bearer tokens. Format of AUTH_DEV_TOKENS:
//
//   token:subject:projA,projB;token2:subject2:projC;token3:sub3:*
//
//   - entries separated by ';'
//   - three fields per entry separated by ':' — token, subject, and a
//     comma-separated project grant list
//   - a project grant of '*' grants every project
//
// Whitespace around entries/fields is trimmed. Malformed entries are
// skipped (logged by the caller if it passes a logger). An unknown token
// → AuthError(401); a known token that does not grant the requested
// project → AuthError(403) at the call site.
import { AuthError, principalFromProjects } from './types';
import type { Authenticator, Principal } from './types';

interface DevTokenGrant {
  subject: string;
  projects: string[];
}

export function parseDevTokens(raw: string): Map<string, DevTokenGrant> {
  const map = new Map<string, DevTokenGrant>();
  for (const entry of raw.split(';')) {
    const trimmed = entry.trim();
    if (!trimmed) continue;
    // Split into at most 3 fields; a project list never contains ':'.
    const firstColon = trimmed.indexOf(':');
    const secondColon = trimmed.indexOf(':', firstColon + 1);
    if (firstColon === -1 || secondColon === -1) continue;
    const token = trimmed.slice(0, firstColon).trim();
    const subject = trimmed.slice(firstColon + 1, secondColon).trim();
    const projectsRaw = trimmed.slice(secondColon + 1).trim();
    if (!token || !subject) continue;
    const projects = projectsRaw
      .split(',')
      .map((p) => p.trim())
      .filter((p) => p.length > 0);
    map.set(token, { subject, projects });
  }
  return map;
}

export class DevAuthenticator implements Authenticator {
  private readonly tokens: Map<string, DevTokenGrant>;

  constructor(rawTokens: string) {
    this.tokens = parseDevTokens(rawTokens);
  }

  /** Number of configured tokens (for boot-time logging). */
  get size(): number {
    return this.tokens.size;
  }

  async authenticate(bearerToken: string): Promise<Principal> {
    const grant = this.tokens.get(bearerToken);
    if (!grant) {
      throw new AuthError(401, 'Unknown or invalid bearer token');
    }
    return principalFromProjects(grant.subject, grant.projects);
  }
}

File: internal/auth/dev.go
package auth

import (
	"context"
	"strings"
)

// devGrant is the parsed grant for a single dev token.
type devGrant struct {
	subject  string
	projects []string
}

// ParseDevTokens parses the AUTH_DEV_TOKENS string into a token→grant map.
//
// Format (entries separated by ';', fields by ':'):
//
//	token:subject:projA,projB;token2:subject2:projC;token3:sub3:*
//
// Whitespace around entries/fields is trimmed; a project grant of '*' grants
// every project. Malformed entries (missing token or subject, fewer than three
// fields) are skipped. Ported from the TS parseDevTokens.
func ParseDevTokens(raw string) map[string]devGrant {
	out := map[string]devGrant{}
	for _, entry := range strings.Split(raw, ";") {
		trimmed := strings.TrimSpace(entry)
		if trimmed == "" {
			continue
		}
		// Split into at most 3 fields; a project list never contains ':'.
		firstColon := strings.IndexByte(trimmed, ':')
		if firstColon == -1 {
			continue
		}
		secondColon := strings.IndexByte(trimmed[firstColon+1:], ':')
		if secondColon == -1 {
			continue
		}
		secondColon += firstColon + 1
		token := strings.TrimSpace(trimmed[:firstColon])
		subject := strings.TrimSpace(trimmed[firstColon+1 : secondColon])
		projectsRaw := strings.TrimSpace(trimmed[secondColon+1:])
		if token == "" || subject == "" {
			continue
		}
		var projects []string
		for _, p := range strings.Split(projectsRaw, ",") {
			if p = strings.TrimSpace(p); p != "" {
				projects = append(projects, p)
			}
		}
		out[token] = devGrant{subject: subject, projects: projects}
	}
	return out
}

// DevAuthenticator implements [Authenticator] over the static AUTH_DEV_TOKENS map.
type DevAuthenticator struct {
	tokens map[string]devGrant
}

// NewDevAuthenticator parses rawTokens and returns a [DevAuthenticator].
func NewDevAuthenticator(rawTokens string) *DevAuthenticator {
	return &DevAuthenticator{tokens: ParseDevTokens(rawTokens)}
}

// Size reports the number of configured tokens (for boot-time logging).
func (d *DevAuthenticator) Size() int { return len(d.tokens) }

// Authenticate resolves a bearer token to a [Principal]. An unknown token
// yields a 401 [*Error].
func (d *DevAuthenticator) Authenticate(_ context.Context, bearerToken string) (Principal, error) {
	grant, ok := d.tokens[bearerToken]
	if !ok {
		return Principal{}, Unauthenticated("Unknown or invalid bearer token")
	}
	return PrincipalFromProjects(grant.subject, grant.projects), nil
}
